### PR TITLE
MULE-16600: Lazy Init: websockets:config is unable to find http:liste…

### PR DESCRIPTION
**Pull request by gsfernandes**
…ner-config

### CODE REVIEW

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java [hunk 1]
```diff
@@ -243,6 +254,40 @@ private void applyLifecycle(List<String> createdComponentModels, boolean applySt
     });
   }
 
+  private BeanDependencyResolver newBeanDependencyResolver() {
```
Comment (lucianoRM) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java]: Does this change for every execution? Wouldn't it be better to have it as a class attribute?

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java [hunk 2]
```diff
@@ -243,6 +254,40 @@ private void applyLifecycle(List<String> createdComponentModels, boolean applySt
...
+    private final BeanDependencyResolver beanDependencyResolver;
+
+    public ComponentConfigurationLifecycleObjectSorter(BeanDependencyResolver beanDependencyResolver) {
+      // Only Object.class type is defined as a single bucket as we just need to sort the configuration components (DSL)
+      super(new Class[] {Object.class});
+      this.beanDependencyResolver = beanDependencyResolver;
+    }
+
+    /**
+     * Implementation that handles duplicates. A->B->C, so if we add B, C will be added but later A will be added and its
```
Comment (lucianoRM) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java]: Not clear.

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java (hunk 1 above)
Comment (gsfernandes) [reply modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java]: Yeap, nice catch.

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java [hunk 3]
```diff
@@ -201,11 +202,25 @@ protected void prepareBeanFactory(ConfigurableListableBeanFactory beanFactory) {
...
   }
 
-  private void applyLifecycle(List<String> createdComponentModels, boolean applyStartPhase) {
+  private void applyLifecycle(Map<String, Object> components, boolean applyStartPhase) {
     muleContext.withLifecycleLock(() -> {
+      // Sorter in order to initialize and start the components according to they dependencies
+      ComponentConfigurationLifecycleObjectSorter componentConfigurationLifecycleObjectSorter =
+          new ComponentConfigurationLifecycleObjectSorter(beanName -> getDependencyResolver()
+              .resolveComponentDependencies(beanName).stream()
+              .map(componentName -> components.get(componentName))
```
Comment (sonar-mulesoft) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java]: Code Smell Code Smell: Replace this lambda with a method reference. ([squid:S1612](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1612))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7761&issues=AWnZSM1wotIXnmsU458-&open=AWnZSM1wotIXnmsU458-)

[ISSUE_KEY]: AWnZSM1wotIXnmsU458-

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java [hunk 4]
```diff
@@ -201,11 +202,25 @@ protected void prepareBeanFactory(ConfigurableListableBeanFactory beanFactory) {
     addBeanPostProcessors(beanFactory, trackingPostProcessor);
   }
 
-  private void applyLifecycle(List<String> createdComponentModels, boolean applyStartPhase) {
+  private void applyLifecycle(Map<String, Object> components, boolean applyStartPhase) {
```
Comment (sonar-mulesoft) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 28 to the 15 allowed. ([squid:S3776](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3776))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7761&issues=AWnZSM1wotIXnmsU4588&open=AWnZSM1wotIXnmsU4588)

[ISSUE_KEY]: AWnZSM1wotIXnmsU4588

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java [hunk 5]
```diff
@@ -201,11 +202,25 @@ protected void prepareBeanFactory(ConfigurableListableBeanFactory beanFactory) {
...
     muleContext.withLifecycleLock(() -> {
+      // Sorter in order to initialize and start the components according to they dependencies
+      ComponentConfigurationLifecycleObjectSorter componentConfigurationLifecycleObjectSorter =
+          new ComponentConfigurationLifecycleObjectSorter(beanName -> getDependencyResolver()
+              .resolveComponentDependencies(beanName).stream()
+              .map(componentName -> components.get(componentName))
+              .filter(component -> component != null)
```
Comment (sonar-mulesoft) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java]: Code Smell Code Smell: Replace this lambda with method reference 'Objects::nonNull'. ([squid:S1612](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1612))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7761&issues=AWnZSM1wotIXnmsU458_&open=AWnZSM1wotIXnmsU458_)

[ISSUE_KEY]: AWnZSM1wotIXnmsU458_

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java [hunk 6]
```diff
@@ -295,14 +337,14 @@ public void initializeComponents(Predicate<org.mule.runtime.config.internal.mode
             .initializeComponents(componentModelPredicate1, applyStartPhase));
   }
 
-  private List<String> createComponents(Optional<Predicate> predicateOptional, Optional<Location> locationOptional,
-                                        Optional<ComponentModelInitializerAdapter> parentComponentModelInitializerAdapter) {
+  private Map<String, Object> createComponents(Optional<Predicate> predicateOptional, Optional<Location> locationOptional,
```
Comment (sonar-mulesoft) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java]: Code Smell Code Smell: Refactor this method to reduce its Cognitive Complexity from 19 to the 15 allowed. ([squid:S3776](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3776))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7761&issues=AWnZSM1wotIXnmsU4589&open=AWnZSM1wotIXnmsU4589)

[ISSUE_KEY]: AWnZSM1wotIXnmsU4589

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java (hunk 3 above)
Comment (lucianoRM) [reply modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java]: Check sonar comment. Method can even be static.

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java (hunk 2 above)
Comment (lucianoRM) [reply modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java]: Maybe change to something like: 
"Implementation that handles duplicates. 
Whenever a new object is added, then all its dependencies are added as well. So if one of those dependencies wants to be added again, it's ignored."

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java [hunk 7]
```diff
@@ -295,14 +337,14 @@ public void initializeComponents(Predicate<org.mule.runtime.config.internal.mode
...
+                                               Optional<ComponentModelInitializerAdapter> parentComponentModelInitializerAdapter) {
```
Comment (lucianoRM) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java]: weird tab.  was this formatted?

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java (hunk 2 above)
Comment (gsfernandes) [reply modules/spring-config/src/main/java/org/mule/runtime/config/internal/LazyMuleArtifactContext.java]: Sure, I forgot to change this javadoc. Makes sense the comment you have suggested.

**Comment by gsfernandes** [comment]
--validate

**Pull request merged**